### DISCUSSION TITLE
Can we talk about "xterm-kitty"? [I don't want to make trouble!]

### DISCUSSION THREAD

**Discussion by egnor** [kovidgoyal/kitty#3873] [General]
I just want to talk! And maybe capture the situation and its ins and outs, for the record. Many many apologies if this was hashed out somewhere (I couldn't find it on quick search), I'll happily delete this if it's redundant or unwelcome.

#### The situation as I understand it

[Per the FAQ](https://sw.kovidgoyal.net/kitty/faq/#i-get-errors-about-the-terminal-being-unknown-or-opening-the-terminal-failing-when-sshing-into-a-different-computer) (and numerous answers to individual questions), it is strongly, strongly recommended to use `term xterm-kitty` (which sets `$TERM` to `xterm-kitty`) when using kitty.

Since `$TERM` is carried through many remote-login protocols (`ssh`, but also the old `telnet`, `rlogin`, etc) this means that every system one plans to log into must be "kitty-ified". Since kitty is not yet especially mainstream, this often means adding `xterm-kitty` to the remote machine's termcap/terminfo database, at least on a per user basis.

To make this convenient, [there is an `ssh` kitten](https://github.com/kovidgoyal/kitty/blob/master/kittens/ssh/main.py) which wraps the `ssh` command with something that replaces the user's command with a shell script that tries to install the `xterm-kitty` terminal definition file before launching whatever the user originally wanted (a specific command or just a login shell).

#### Consequences and issues

For 99% of users, this works 99% of the time.

However, the ssh kitten makes a lot of assumptions about the destination machine:
- That it's vaguely Unix-like
- That `/bin/sh` and various other tools work in more or less standard ways
- That it has a termcap/terminfo type system
- That it's possible and acceptable to modify the user's home directory

These assumptions are decidedly _not_ true all of the time. It's not at all uncommon to log into non-Unix systems with `ssh` (embedded systems, console drivers, other operating systems, etc). It's not at all uncommon to log in as captive users without write permissions. I don't know how common these use cases are, but there are plenty of threads about various aspects of this problem.

More philosophically, the ssh-wrapper approach feels like a brittle solution with significant possibility of unanticipated and surprising outcomes.

#### Why?

Why insist on using `xterm-kitty` instead of something more widely recognized like `xterm-256color`? The reasons aren't documented that I can find, but I assume the intent is to control the terminfo definition so that it precisely matches what kitty implements. Furthermore it may allow kitty-aware tools (such as the remote controller and anything which uses kitty-specific extensions) to detect kitty's presence or absence and behave accordingly.

I'm very curious for a deeper explanation here, including the specific definitions that need to get nailed down. I'm intimately familiar with terminfo/termcap and Unix terminal handling in general, and can receive arcane specifics without needing explanation. I'm happy to write a FAQ appendix as needed...

#### Alternatives?

Since I am very definitively in the 1% not well served by an ssh-wrapper, I would love to set `$TERM` to something more widely understood than `xterm-kitty`. (I usually find `xterm-256color` a good compromise.)

As Kovid warns, this does seem to break some things, e.g. https://github.com/kovidgoyal/kitty/issues/3870. ("If you change TERM everything will break" was the reply; "everything" is surely hyperbole, but at least some things do break!) It seems possible to remedy the problems I've seen, but I definitely don't understand all the issues/

To be clear, terminfo/termcap is such an old and weird system that I don't think there's an architecturally _right_ way to solve any related issue. The whole idea of passing a `$TERM` value around without passing a definition is asking for trouble, and effectively leads to everyone settling on some commonly understood core, and occasionally people extending it (as in the infamous `rxvt-unicode-256color`), battling to get some new entry added everywhere, and declaring success at the 95% mark. Furthermore these days many applications assume xterm-like terminal behavior anyway.

Expanding or extending `ssh` across the world would be an immense task (there are many implementations), and `ssh` isn't the only tool that carries `$TERM` values from system to system. Getting `xterm-kitty` onto every system in the universe is also implausible.

IMHO, it _seems_ better just to relegate terminfo/termcap to the past, emulate xterm faithfully, set `$TERM` to some widely understood value, and use other mechanisms (such as live detection) for apps to detect when specific extensions are available. BUT, I say this not knowing the probably very good reasons why Kovid has gone to great lengths to try to keep everyone on `xterm-kitty`.

**Comment by kovidgoyal** [comment]
On Mon, Jul 26, 2021 at 09:14:17PM -0700, Daniel Egnor wrote:
> However, the ssh kitten makes a lot of assumptions about the destination machine:
> - That it's vaguely Unix-like

Does *interactive* SSH actually work on non Unix like systems?

> - That `/bin/sh` and various other tools work in more or less standard ways

The ssh kitten relies on POSIX tools only, as far as I know. If you are
trying to log into non POSIX systems, you are on your own.

> - That it has a termcap/terminfo type system

What system other than termcap/terminfo uses the TERM variable?

> - That it's possible and acceptable to modify the user's home directory

If you are not going to modify the home directory at all, what is it you
are doing on the remote server in an interactive session? Modifying
files belonging to someone else? Simply the act of logging into an
interactive session will modify files in your home directory (the shell
will write history/login info), sshd will write to its log files, etc.
So I dont see what's so egregious about writing one file into
~/.terminfo

> 
> These assumptions are decidedly _not_ true all of the time. It's not at all uncommon to log into non-Unix systems with `ssh` (embedded systems, console drivers, other operating systems, etc). It's not at all uncommon to log in as captive users without write permissions. I don't know how common these use cases are, but there are plenty of threads about various aspects of this problem.

See above. 

> 
> More philosophically, the ssh-wrapper approach feels like a brittle solution with significant possibility of unanticipated and surprising outcomes.

I cant think of any unanticipated or surprising outcome, other than, the
terminfo database failing to be copied as a result of which things
will not work. Which is the same thing you risk with setting TERM to
whatever. Feel free to enlighten me.

The medium term robust solution is for the ssh daemon to copy the
terminfo database onto the remote server when running an interactive
session. 

> IMHO, it _seems_ better just to relegate terminfo/termcap to the past, emulate xterm faithfully, set `$TERM` to some widely understood value, and use other mechanisms (such as live detection) for apps to detect when specific extensions are available. BUT, I say this not knowing the probably very good reasons why Kovid has gone to great lengths to try to keep everyone on `xterm-kitty`.

The very good reason is that live detection is *slow* and *complex*. It
involves a round trip to the terminal emulator. It is *eminently* not
suited for quick invocation command line tools. Therefore, one needs
some form of terminal capabilities database. Sadly the one we are stuck
with is terminfo. Efforts to replace it have failed:
https://gitlab.freedesktop.org/terminal-wg/specifications/-/issues/8

So we are stuck with terminfo. And terminfo's maintainer is, to be
polite, less than ideal. See #879 or
https://gist.github.com/XVilka/8346728#detection 

Since kitty aims to move the terminal ecosystem forward, it cannot
have the only way to detect its capabilities without a roundtrip be tied
to the boat anchor that is xterm and ncurses.

**Comment by egnor** [comment]
To be clear, I'm not trying to pick a fight, I super love what you're doing with kitty, I'm just poking a bit at this particular edge. I hadn't bumped into https://github.com/kovidgoyal/kitty/issues/879 which is interesting reading for sure.

> Does *interactive* SSH actually work on non Unix like systems?

> If you are not going to modify the home directory at all, what is it you are doing on the remote server in an interactive session?

> I cant think of any unanticipated or surprising outcome, other than, the terminfo database failing to be copied as a result of which things will not work. Which is the same thing you risk with setting TERM to whatever. Feel free to enlighten me.

Again, I think it all works fine for 99% of users 99% of the time, but here's some of what's in the 1%:

- Cisco switches (IOS), IBM mainframes (z/OS), embedded RTOS (VxWorks, etc), console servers (Opengear, etc) all support interactive ssh logins and care about terminal type to some degree. Some of them ignore the ssh terminal type entirely (no problem!), others have some set of accepted types that includes things like `vt100` and `xterm` but certainly not `xterm-kitty`. And I sure as heck don't want to run a bunch of Bourne-shell-intended code on a Cisco router. (_Hopefully_ I'd just get a syntax error. Admittedly, it's been many years since I was logging into those on the regular.)

- Stripped-down mini Unix-based systems often have static terminfo files and no `tic` type command. You've gotten questions about those, your answer is to copy things over manually, which can work if the home directory is writable (by no means a given, they often have no/little writable filesystem) but is obviously a hassle if it's a thing one is doing a lot.

- Sysadmins of large cluster/cloud operations often log in as users that have no home dir (or a read only home dir), and then gain specific privileges (`sudo` or similar) to take any actions. When you have hundreds or thousands of machines it's important not to let the state of individual machines wander (otherwise things get unpredictable) hence the lockdown.

- Firewall/VPN gateway servers often have a locked down account users "bounce" through, first you log into the gateway and then you run another ssh (or other) command to get where you're going. The account on the gateway has no writable home dir because it's a bounce pad shared by lots of users.

Again these are arguably obscure cases and tend to be used by more skilled users who can find workarounds. In fact I am one of those relatively skilled users (on a good day, anyway). But I'd rather I didn't have to find workarounds. Most importantly I need to know what will break when `$TERM` is not `xterm-kitty`. The current messaging reads as "everything will break! don't do it!" which isn't as helpful as knowing "ok, kitty-specific utilities like `icat` and remote control won't work, but nothing will explode". Or maybe something will explode? I was surprised by that kitty shell issue!

> Since kitty aims to move the terminal ecosystem forward, it cannot have the only way to detect its capabilities without a roundtrip be tied to the boat anchor that is xterm and ncurses.

And I admire that attitude for sure! Doing anything innovative in this space always requires picking some careful battles about which ancient stone pyramids to carefully preserve which which to bulldoze. (So much about this evolved ecosystem is ludicrous...)

Anyway, what I'm hearing is that the main issue is detecting kitty-specific capabilities (above and beyond the generic terminal types) without the delay of a probing round-trip. In that case, it _seems_ like setting `$TERM` to some "generic" value like `xterm` or `xterm-256color` or whatever could be documented (and debugged) as "you'll lose kitty goodies, but everything else should work". Am I wrong?

My super humble modest proposal:

Leave `$TERM` and terminfo to rot. Use a second environment variable to describe terminal extensions (anything above and beyond ECMA-48). Call it `$LC_TERM_ABILITIES`. Instead of the terminfo format it would just be a colon delimited list of functionality names, including well known things like `color8bit` or `color24bit` and also namespaced/experimental things like `net.kovidgoyal.kitty` (or `net.kovidgoyal.kitty.graphics` if you want to be more granular). Programs would use the ones they understand and ignore the ones they don't so adding extensions is never a problem.

A lot of distros already pass and accept `$LC_*` variables over ssh so you're almost done (this is a hack but a tight one). No need to wade through a committee, just do it and tell your friends who enjoy pretty terminal things. Get the node folks (the biggest fans of fancy terminal eyebleed I've met) on board and you're golden. And sure, offer a kitten that tunnels `$LC_TERM_ABILITIES` by wrapping `ssh`, for cases where the variable isn't already passed and people miss the extra capabilities on the remote end and they're not worried about obscure cases. Boom! Everything is great and nobody is complaining.

(I'm sure there are Many Issues with this...)

**Reply by 0cjs** [reply]
> Again, I think it all works fine for 99% of users 99% of the time, but here's some of what's in the 1%:

You missed a big one: shared terminal sessions.

I do a lot of remote pair programming where we share terminals with [tmate](https://tmate.io). (I also know people who share with Gnu screen.) Whether I'm using kitty or not, the other folks in the session are not using it. I would not be surprised if I've never been in a session where even two members are using the same terminal program; I'm pretty certain I've never been in one where everybody is.

This is where the "not compatible with xterm" stuff worries me: if it's a situation where everybody must use kitty or nobody, it's definitely going to be nobody.

**Comment by kovidgoyal** [comment]
On Tue, Jul 27, 2021 at 01:34:49AM -0700, Daniel Egnor wrote:
> To be clear, I'm not trying to pick a fight, I super love what you're doing with kitty, I'm just poking a bit at this particular edge. I hadn't bumped into https://github.com/kovidgoyal/kitty/issues/879 which is interesting reading for sure.
> 
> > Does *interactive* SSH actually work on non Unix like systems?
> 
> > If you are not going to modify the home directory at all, what is it you are doing on the remote server in an interactive session?
> 
> > I cant think of any unanticipated or surprising outcome, other than, the terminfo database failing to be copied as a result of which things will not work. Which is the same thing you risk with setting TERM to whatever. Feel free to enlighten me.
> 
> Again, I think it all works fine for 99% of users 99% of the time, but here's some of what's in the 1%:
> 
> - Cisco switches (IOS), IBM mainframes (z/OS), embedded RTOS (VxWorks, etc), console servers (Opengear, etc) all support interactive ssh logins and care about terminal type to some degree. Some of them ignore the ssh terminal type entirely (no problem!), others have some set of accepted types that includes things like `vt100` and `xterm` but certainly not `xterm-kitty`. And I sure as heck don't want to run a bunch of Bourne-shell-intended code on a Cisco router. (_Hopefully_ I'd just get a syntax error. Admittedly, it's been many years since I was logging into those on the regular.)

All of these run some flavor of UNIX, so I dont see the problem.

It's not bourne shell, it's POSIX. And if you dont want to run anything
on the server (a completely illogical attitude to me, given that you
will be running an entire shell, but that's your choice), write a five
line script that copies the compiled terminfo db and then sshes in. If
you want to be thorough about it, use a ssh controlmaster so there is no
performance penalty either.

> 
> - Stripped-down mini Unix-based systems often have static terminfo files and no `tic` type command. You've gotten questions about those, your answer is to copy things over manually, which can work if the home directory is writable (by no means a given, they often have no/little writable filesystem) but is obviously a hassle if it's a thing one is doing a lot.

Same as above.

> 
> - Sysadmins of large cluster/cloud operations often log in as users that have no home dir (or a read only home dir), and then gain specific privileges (`sudo` or similar) to take any actions. When you have hundreds or thousands of machines it's important not to let the state of individual machines wander (otherwise things get unpredictable) hence the lockdown.

A terminfo db in ~/.terminfo is not server state. It is never used by
anything in the absence of TERM=xterm-kitty which nothing on the server
will set, other than SSH, when it is needed.

> 
> - Firewall/VPN gateway servers often have a locked down account users "bounce" through, first you log into the gateway and then you run another ssh (or other) command to get where you're going. The account on the gateway has no writable home dir because it's a bounce pad shared by lots of users.

I assume you mean using ssh -J. This should work perfectly transparently
with the ssh kitten, though I haven't actually tested it.

> 
> Again these are arguably obscure cases and tend to be used by more skilled users who can find workarounds. In fact I am one of those relatively skilled users (on a good day, anyway). But I'd rather I didn't have to find workarounds. Most importantly I need to know what will break when `$TERM` is not `xterm-kitty`. The current messaging reads as "everything will break! don't do it!" which isn't as helpful as knowing "ok, kitty-specific utilities like `icat` and remote control won't work, but nothing will explode". Or maybe something will explode? I was surprised by that kitty shell issue!

They are indeed edge cases and ones that have perfectly serviceable
workarounds in about five minutes of scripting. terminfo is used for
everything from keypresses (the cause of readline and hence kitty shell
breakage), to how to erase the screen or a line, to how to reset the
terminal to how to switch to alternate modes, etc, etc. Just read man
terminfo. With a mismatched terminfo literally *anything* can break.

> 
> > Since kitty aims to move the terminal ecosystem forward, it cannot have the only way to detect its capabilities without a roundtrip be tied to the boat anchor that is xterm and ncurses.
> 
> And I admire that attitude for sure! Doing anything innovative in this space always requires picking some careful battles about which ancient stone pyramids to carefully preserve which which to bulldoze. (So much about this evolved ecosystem is ludicrous...)
> 
> Anyway, what I'm hearing is that the main issue is detecting kitty-specific capabilities (above and beyond the generic terminal types) without the delay of a probing round-trip. In that case, it _seems_ like setting `$TERM` to some "generic" value like `xterm` or `xterm-256color` or whatever could be documented (and debugged) as "you'll lose kitty goodies, but everything else should work". Am I wrong?

No, the main issue is I need a way to be able to add capabilities
to kitty's terminfo database without being gatekeepered by xterm.
That is important because querying the terminal is not acceptable
in too many contexts.

> 
> My super humble modest proposal:
> 
> Leave `$TERM` and terminfo to rot. Use a second environment variable to describe terminal extensions (anything above and beyond ECMA-48). Call it `$LC_TERM_ABILITIES`. Instead of the terminfo format it would just be a colon delimited list of functionality names, including well known things like `color8bit` or `color24bit` and also namespaced/experimental things like `net.kovidgoyal.kitty` (or `net.kovidgoyal.kitty.graphics` if you want to be more granular). Programs would use the ones they understand and ignore the ones they don't so adding extensions is never a problem.
>
> 
> A lot of distros already pass and accept `$LC_*` variables over ssh so you're almost done (this is a hack but a tight one). No need to wade through a committee, just do it and tell your friends who enjoy pretty terminal things. Get the node folks (the biggest fans of fancy terminal eyebleed I've met) on board and you're golden. And sure, offer a kitten that tunnels `$LC_TERM_ABILITIES` by wrapping `ssh`, for cases where the variable isn't already passed and people miss the extra capabilities on the remote end and they're not worried about obscure cases. Boom! Everything is great and nobody is complaining.
> 
> (I'm sure there are Many Issues with this...)

You say that getting the kitty terminfo onto all machines is impossible,
based on a bunch of edge cases. And yet you seem to think that I can
wave a magic wand and get an env var through all those edge cases. I can
guarantee that your various stripped down boxen will not pass through
LC_*. Hell, even my laptop based on Arch doesnt::

    LC_NO_CHANCE=1 ssh localhost env | grep LC

**Comment by egnor** [comment]
[shameless buttering 🧈 - reminder that I find Kitty is incredibly great, fast, flexible and thoughtfully designed, a reminder of the golden era of desktop creativity and hopefully a harbinger of new greatness 😻]

My takeaway is: Those of us who (for whatever reason) run with `$TERM` != `xterm-kitty` are doing something "unsupported". You've given us a kitty config setting, which we appreciate, and if we find stuff that's broken, we're free to send you PR's, but you're not going to spend time debugging it, because you're just not interested in that particular compatibility path; you'd rather spend the energy finding ways to get `xterm-kitty` recognized everywhere, ideally in an automatically distributed way; you feel like that's the best path for the ecosystem in general.

(And if we spend you PRs to clarify your FAQ entries a bit, you'll consider them, but no guarantees!)

Sound about right?

Back into the trenches below, feel free to ignore if this has already eaten too much of your time.

> terminfo is used for everything from keypresses (the cause of readline and hence kitty shell breakage), to how to erase the screen or a line, to how to reset the terminal to how to switch to alternate modes, etc, etc. Just read man terminfo. With a mismatched terminfo literally *anything* can break.

(I am intimately familiar with terminfo, sadly...)

You put a ton of effort, which I appreciate, into making kitty backward-compatible with xterm at the escape code level (ignoring weird nonsense like Tektronix graphics). (I'm pretty sure it's a ton of effort because I've written terminal emulators and know it's a pain to get all the edge cases right.)

So, `$TERM=xterm` really should work, given any terminfo/termcap definition that works with xterm itself. This is not an arbitrarily mismatched terminfo. (Of course if you set `$TERM=vt52` kitty won't work, neither will xterm, but nobody's doing that.)

With `$TERM=xterm` you wouldn't get fancy extensions that allow displaying images and whatnot, but at least basic compatibility should be there. Notably, readline-using programs work just fine under kitty with `$TERM=xterm`, the inbuilt kitty shell must be doing something different. I'll try to figure out what it is. (It sure looks like it just uses vanilla readline in a straightforward way, so it's a bit of a mystery.)

> You say that getting the kitty terminfo onto all machines is impossible, based on a bunch of edge cases. And yet you seem to think that I can wave a magic wand and get an env var through all those edge cases. I can guarantee that your various stripped down boxen will not pass through LC_*. Hell, even my laptop based on Arch doesnt.

The difference is in graceful fallback and backward compatibility.

If `$TERM=xterm` and `$LC_TERM_ABILITIES` (or whatever "second path") is lost or ignored, you end up with vanilla xterm compatibility, which should work just fine, even if you don't have as many fancy colors and image display options. (Those obscure systems couldn't use that anyway.) That feels like a good, graceful, backward compatible fallback. (And, passing `$LC_TERM_ABILITIES` is an easy, robust ssh config option.)

If `$TERM=xterm-kitty` and the destination doesn't understand `xterm-kitty`, then you get error messages and don't have terminal support at all. (And if you use the `ssh` kitten and the destination doesn't speak POSIX shell syntax, then you get error messages and no login at all.) That seems like an unfortunate incompatibility.

It's a tradeoff between sometimes losing capability vs sometimes failing entirely (with different rates of each). I like the graceful fallback option myself but I can see different viewpoints here.

The `$LC_TERM_ABILITIES` thing was just a thought (I think I like terminfo a lot less than you do, I'm happy if it dies a slow death). But even if the only way to detect kitty extensions is via terminfo and `$TERM=xterm-kitty`, I think it's reasonable to expect `$TERM=xterm` to have basic functionality working (and for the most part it does work just fine! -- I just bumped into that shell thing).

🕊️

**Comment by kovidgoyal** [comment]
On Tue, Jul 27, 2021 at 11:39:02AM -0700, Daniel Egnor wrote:
> My takeaway is: Those of us who (for whatever reason) run with `$TERM` != `xterm-kitty` are doing something "unsupported". You've given us a kitty config setting, which we appreciate, and if we find stuff that's broken, we're free to send you PR's, but you're not going to spend time debugging it, because you're just not interested in that particular compatibility path; you'd rather spend the energy finding ways to get `xterm-kitty` recognized everywhere, ideally in an automatically distributed way; you feel like that's the best path for the ecosystem in general.
> 
> (And if we spend you PRs to clarify your FAQ entries a bit, you'll consider them, but no guarantees!)
> 
> Sound about right?

You are free to send PRs for pretty much anything, I will accept or
reject them based on their individual merits. If you think the topic is
likely to be controversial, open an issue first to discuss it, to avoid
wasting your time.

I am indeed not going to spend time on TERM != xterm-kitty issues.

> 
> Back into the trenches below, feel free to ignore if this has already eaten too much of your time.
> 
> > terminfo is used for everything from keypresses (the cause of readline and hence kitty shell breakage), to how to erase the screen or a line, to how to reset the terminal to how to switch to alternate modes, etc, etc. Just read man terminfo. With a mismatched terminfo literally *anything* can break.
> 
> (I am intimately familiar with terminfo, sadly...)
> 
> You put a ton of effort, which I appreciate, into making kitty backward-compatible with xterm at the escape code level (ignoring weird nonsense like Tektronix graphics). (I'm pretty sure it's a ton of effort because I've written terminal emulators and know it's a pain to get all the edge cases right.)

A pain and a half!

> 
> So, `$TERM=xterm` really should work, given any terminfo/termcap definition that works with xterm itself. This is not an arbitrarily mismatched terminfo. (Of course if you set `$TERM=vt52` kitty won't work, neither will xterm, but nobody's doing that.)

Not quite, there are some places I intentionally diverge from xterm,
most notably bce and few of the more exotic keys. You can check for
yourself easily using infocmp to diff terminfo.

> 
> With `$TERM=xterm` you wouldn't get fancy extensions that allow displaying images and whatnot, but at least basic compatibility should be there. Notably, readline-using programs work just fine under kitty with `$TERM=xterm`, the inbuilt kitty shell must be doing something different. I'll try to figure out what it is. (It sure looks like it just uses vanilla readline in a straightforward way, so it's a bit of a mystery.)

I will note that 

kitty --config NONE -o term=xterm

has kitty shell working fine for me.

> 
> > You say that getting the kitty terminfo onto all machines is impossible, based on a bunch of edge cases. And yet you seem to think that I can wave a magic wand and get an env var through all those edge cases. I can guarantee that your various stripped down boxen will not pass through LC_*. Hell, even my laptop based on Arch doesnt.
> 
> The difference is in graceful fallback and backward compatibility.
> 
> If `$TERM=xterm` and `$LC_TERM_ABILITIES` (or whatever "second path") is lost or ignored, you end up with vanilla xterm compatibility, which should work just fine, even if you don't have as many fancy colors and image display options. (Those obscure systems couldn't use that anyway.) That feels like a good, graceful, backward compatible fallback. (And, passing `$LC_TERM_ABILITIES` is an easy, robust ssh config option.)

The thing you are missing is that if the path is extremely unreliable,
as LC_* would be, the larger ecosystem wont use it for detection. So we
need some mechanism that is relatively reliable. And terminfo, with the
ssh kitten, is the best I can come up with without support from either
kernel developers or openssh developers. 

> 
> If `$TERM=xterm-kitty` and the destination doesn't understand `xterm-kitty`, then you get error messages and don't have terminal support at all. (And if you use the `ssh` kitten and the destination doesn't speak POSIX shell syntax, then you get error messages and no login at all.) That seems like an unfortunate incompatibility.

Note the destination can use whatever shell it likes, what the ssh
kitten requires is that */bin/sh* be POSIX compatible. That's a pretty
basic requirement, which if your system does not follow, lots of other
things will break.

**Comment by ewhac** [comment]
I found this discussion because I've been bumping in to the same problem when logging in to remote systems that have a rather pared-down terminfo database.  My workaround has been to set `$TERM` to `rxvt`, which seems to work well enough to allow `less` and `vi` to work without major issue.

Since we can't expect `xterm-kitty` to always exist on the remote system, I'm wondering if it might be reasonable to add a command line option or environment variable that launches kitty into a somewhat dumbed-down compatibility mode.  That mode could be `xterm`, `vt100`, `ansi`, `wyse50`, whatever, so long as it's commonly available/understood on remote systems.

**Reply by egnor** [reply]
Do you find that `vt100` doesn't work? I'd expect it to work fine, just be missing color and other such attributes.

**Reply by josebalius** [reply]
@ewhac thank you so much. Your suggestion here saved the day for me `rxvt`.

**Comment by pnunn** [comment]
I know this is an OLD thread, but I've just run into this using teleport.

When I use tsh ssh server_name to log in, I get the issue of the terminal being set to xterm-kitty and things get funky in my fish session on the remote server.

Pretty sure the wrapper isn't going to help me here because I have no way to run it over the teleport connection as far as I know.

Is there a way to update the remote system without using the wrapper that would help here?

**Comment by kovidgoyal** [comment]
On Mon, Sep 25, 2023 at 05:37:39PM -0700, Peter Nunn wrote:

> Is there a way to update the remote system without using the wrapper that would help here?

This is documented in the FAQ.
https://sw.kovidgoyal.net/kitty/faq/#i-get-errors-about-the-terminal-being-unknown-or-opening-the-terminal-failing-or-functional-keys-like-arrow-keys-don-t-work

https://sw.kovidgoyal.net/kitty/kittens/ssh/#manual-terminfo-copy

**Reply by pnunn** [reply]
Thank you. I missed the link to the manual copy. I'll give it a go.

**Comment by ismell** [comment]
While reading `man termcap` I found the following:

> The  TERMINFO variable can contain a terminal description instead of the pathname of a terminal database.  If this variable begins with “hex:” or “b64:” then ncurses reads a terminal description from hexadeci‐
> mal- or base64-encoded data, and if that description matches the name sought, will use that.  This encoded data can be set using the “-Q” option of tic or infocmp.


Could kitty add an option to pass the contents of `xterm-kitty` in the `TERMINFO` variable?

**Reply by ismell** [reply]
FWIW, I tested passing in TERMINFO into a container environment that doesn't have kitty's terminfo installed, and it works like a charm! See https://chromium-review.googlesource.com/c/chromiumos/bazel/+/5383676.

**Reply by ewhac** [reply]
> FWIW, I tested passing in TERMINFO into a container environment that doesn't have kitty's terminfo installed, and it works like a charm! See https://chromium-review.googlesource.com/c/chromiumos/bazel/+/5383676.

Tried this on a local server, and it appears to work.

Unfortunately, `sshd`'s default configuration accepts no environment variables from the client (except `TERM` when a pseudo-terminal is requested)  Debian's `sshd` package is slightly more liberal, allowing `COLORTERM`, `NO_COLOR`, and locale variables through.  `sshd_config` must be modified to accept the `TERMINFO` environment variable by adding the line:
```
AcceptEnv TERMINFO
```
Which may be a non-starter in some environments/organizations.

**Comment by kovidgoyal** [comment]
On Wed, Mar 20, 2024 at 12:41:18PM -0700, Raul E Rangel wrote:
> Could kitty add an option to pass the contents of `xterm-kitty` in the `TERMINFO` variable?

I can certainly add that, but note that this is likely supported only by
ncurses itself. Other libraries/software will probably fail with
TERMINFO not set to a path.

**Reply by kovidgoyal** [reply]
And here you go: https://github.com/kovidgoyal/kitty/commit/198b69e275b8f5db6eff5e71f02e86c8ae4e7b8c

**Comment by IvoARodrigues** [comment]
For ssh usage, I find useful to use the `~/.ssh/config` and set:

```
Host *
    SetEnv TERM=xterm-256color
```

**Reply by rosejn** [reply]
oh thank god!  This is what I've been needing.

@kovidgoyal seems particularly defensive about this damn kitten, but I am constantly running into issues with the kitty terminal setup over SSH.  We tend to use scripts to SSH into cloud instances where we are pulling keys from vault, loading into ssh-agent, and then logging into the machine, and even if I create a wrapper script for ssh which calls "kitten ssh" under the hood it just hangs.  I think this should be the default for remote kitty terminals, and the xterm-kitty should only be used when explicitly using the ridiculous kitten command.

**Reply by stronk-dev** [reply]
Wouldn't it be better to use `TERM=kitty` in that case? That should be recognized and should be more compatible.

But yeah, really annoying that it's not compatible out of the box. `kitten` doesn't even work when SSH'ing as `root` into a system (at least not for me). I would be fine with using `kitten` if it were to function as if I was using `SSH` from a 'normal' terminal.
For me the only solution is override `$TERM` (which breaks some things) or install the `kitty-terminfo` package (which really unaceptable when dealing with remote production machines).

**Comment by erlangparasu** [comment]
`htop`

`Error opening terminal: xterm-kitty.`

**Comment by tbe** [comment]
After switching to kitty, i land up with the same issues.

And there are still some missconceptions:

a) WINDOWS has sshd support. Interactive sessions are working there, and it is a non-Unix, non-Posix system
b) High Security Environments are a thing! Read-Only homedirs for example, as mentioned. 
c) Intercepting security systems. There are systems in use, that intercept SSH connections and limit what you can do. Copying files to servers, no matter the way, is prohibited in many environments
d) This is more then a simple wrapper. There are thousands of code lines just to build up an SSH connection. This violates for example my company policies. I only may use our PAM client and the original, distributed provided, SSH binary.

"Works for me" is a very bad idea here. Currently, i will test what happens if i just force xterm-265color in the SSH config.

At least this issue should be documented very prominently. It is just fucked up to build the perfect kitty, just to throw it out if you have to do REAL work with it.

**Comment by abatkin** [comment]
Yeah there seems to be some cognitive dissonance going on here.

It doesn't matter how hard people argue that systems should support xterm-kitty or there are wrapper scripts or workarounds or whatever. There are very common workflows where that DOES NOT WORK and that CANNOT CHANGE. It's actually incredibly common.

In fact, in my daily workflow, 95% of my systems work just fine, including the ones where I spend the majority of my time. But that 5% that I can't control of affect are so painful, that I can't use Kitty as my daily driver anymore, and that's disappointing.

I think in my perfect world, Kitty would "officially support" running in xterm-256color or xterm mode _with reduced functionality_ and then we can pick and choose what TERM to set on different systems.

**Comment by ksdev-pl** [comment]
Dear future reader, set the term to ⁠xterm-256color in Kitty configuration to preserve your sanity, even if it means losing some Kitty-specific features. If basic terminal functionality ever breaks when using ⁠xterm-256color, save yourself the headache and switch to a different terminal.

**Reply by tbe** [reply]
Or, just switch to wezterm now. Should cover most of the aspects of kitty + integrated search and more without 3rd-party tools.

**Comment by MelvinQc542** [comment]
Although I am not an expert like most of you guys seem to be, I wish to add my grain of salt just for my tiny-easy use case.  I simply want to remotely log in my MacBook from my arch linux system.

Before I switched to nix flakes on my Mac to deterministically configure my machine, all was Ok - and xterm-kitty was not causing any problem.  With my flake, though (and I will find a solution for certain), the xterm-kitty emulation was not recognized - therefore, I have to run `TERM=xterm-256color ssh macbookpro` for my backspace and Ctrl-L to keep working.

Maybe this "not-so-technical" insight could help to reach a resolution.  ;)

BTW, @kovidgoyal, I have just found kitty and I loooove it!  I use it on all my machines.  THANKS!!

**Comment by rmmh** [comment]
Would a patch to enable `bce` (background color erase) potentially be accepted? That seems to be the most impactful difference shown in `infocmp -d xterm-256color xterm-kitty`.

**Comment by kovidgoyal** [comment]
On Tue, Dec 02, 2025 at 09:50:38AM -0800, Ryan Hitchman wrote:
> Would a patch to enable `bce` (background color erase) potentially be accepted? That seems to be the most impactful difference shown in `infocmp -d xterm-256color xterm-kitty`.

No, bce is an anti feature as described in exhaustive detail
at the links pointed to in https://sw.kovidgoyal.net/kitty/deccara/

**Comment by robinchrist** [comment]
So, I've just stumbled over Kitty and I'd love to try it out, looks very promising.

What worries me is the entire SSH / terminfo thing. I'd mainly use Kitty for work on remote machines via SSH.

The problem is: I work as a Network Engineer. That means my remote machines are basically all "appliance style", and mostly I will not be able to modify the home directory.

Maybe that's a new POV for the problem - although it's a bit of a niche use case :)

**Reply by MelvinQc542** [reply]
I’ve actually moved to WezTerm.  The issues I described in my post did not materialize.

**Reply by lady-melinoe** [reply]
since openssh 8.7 (i think), you can add something like
```
Host server.my.domain
	SetEnv TERM=xterm-256color
```
to your .ssh/config and then just run `ssh server.my.domain` 
(there is also quite a few other cool options in there, see `man ssh_config` )
